MILESTONE-1.md

```diff
@@ -0,0 +1,101 @@
+# Milestone 1
+
+The first milestone includes 3 contracts (in red on the diagram below): A space contract holding votes and proposals, an authenticator contract approving signed messages for casting a vote or creating a proposal, and a voting strategy where 1 address is equal to 1 voting power. Execution, custom settings, or storage proof calculations are not part of this milestone.
+
+![](https://user-images.githubusercontent.com/16245250/155883647-8f853411-bea9-4a73-bd69-2be850d4ffc9.png)
+
+Source: https://whimsical.com/snapshot-x-design-proposal-Gy1y3P6zTs4rdE4cStaKW4
+
+## Requirements
+
+Here are the actions to cover in this milestone:
+- Use auth contract as proxy
+- Create a proposal
+- Cast a vote
+- Store votes and proposals
+
+## Interfaces
+
+### Space contract
+
+This is the main contract, previously called "voting contract". The contract is responsible for storing proposals and votes, it also defines and stores parameters for the space.
+
+#### Storage vars
+
+- **voting_delay** = 600   
+The time in seconds to wait between when a proposal is created and when the proposal voting period starts. 
+
+- **voting_period** = 3600   
+The duration of a proposal in seconds.
+
+- **proposal_threshold** = 1   
+The minimum amount of voting power that is required to submit a proposal.
+
+- **voting_strategy** = 0x123...   
+The voting strategy contract address that is used to calculate the voting power for each voter, the strategy will return 1 voting power for any address. 
+
+- **auth** = 0x123...  
+The authenticator contract address that is used to verify the addresses of voters and proposers.
+
+- **next_proposal_nonce** = 1  
+An increasing nonce to make the proposal id unique.
+
+- **proposals** = TBD  
+A mapping of proposals by ids with required information: execution_hash, start, end.
+
+- **votes** = TBD  
+A mapping of votes by proposal id with required information: address, choice, and voting power.
```

Comment:
I think we can avoid this mapping and save storage_vars by doing the following:

1. Adding a mapping called `has_voted` that maps between a pair of voting_address, proposal_id into a flag that says whether the address already voted on the proposal
2. Adding three mappings that map between proposal_id to a counter. One for the sum of "Yes" votes. One for the sum of "No: votes, and one for the sum of "Abstain" votes

This way we have only two writes for each vote. Instead of 3 here and at least one more to manage the array of votes per proposal.

Reply:
The only problem I see is that we can't query any storage var that would show us the list of voters for a specific `proposal_id`. Don't know if that's an issue or not :)

Reply:
Also @ShahakShama , do you think we could get away with `votes(proposal_id: felt, choice: felt) -> (count: felt)` ?

This would remove a big `if else if else if else` chunk of ugly code :p

Reply:
for reference: I implemented it in another branch if we ever decide to switch to it:

https://github.com/snapshot-labs/sx-core/commit/7b2060fe034945aa98ea351fe21f4cce7ed229bb

Reply:
> The only problem I see is that we can't query any storage var that would show us the list of voters for a specific `proposal_id`. Don't know if that's an issue or not :) (maybe @bonustrack ? :)

@psott It won't be an issue we can emit an event and get the information with the API

Reply:
@pscott Great Idea!